File: codes_trial/rust/007.Functions.rs
#![allow(unused_variables)]
#![allow(unused_parens)]
#![allow(dead_code)]
#![allow(unused_must_use)]



// ? < function statement or function declaration

fn get_rectangle_area_v1(rectangle_width: i64, rectangle_length: i64) -> i64 {
    return rectangle_width * rectangle_length;
}

// ? function statement or function declaration />



// ? < Passing functions as arguments to other functions

fn say_hello_v1<F>(callback_function: F) where F: Fn() {
    println!("hello");
    callback_function();
}

fn say_how_are_you_v1() {
    println!("how are you?");
}

// ? Passing functions as arguments to other functions />



// ? < Returning functions as values from other functions

fn multiply_v1(a: i64) -> impl Fn(i64) -> i64 {
    return move |b| a * b;
}

fn multiply_v2(a: i64) -> Box<dyn Fn(i64) -> i64> {
    return Box::new(move |b| a * b);
}

// ? Returning functions as values from other functions />



fn main() {
    // ? < function expression with closure syntax or named closure syntax function

    let get_rectangle_area_v2 = |rectangle_width: i64, rectangle_length: i64| -> i64 {
        return rectangle_width * rectangle_length;
    };

    let get_rectangle_area_v3 = |rectangle_width: i64, rectangle_length: i64| rectangle_width * rectangle_length;

    // ? function expression with closure syntax or named closure syntax function />



    // ? < anonymous function expression with closure syntax
    
    |rectangle_width: i64, rectangle_length: i64| -> i64 {
        return rectangle_width * rectangle_length;
    };
    
    |rectangle_width: i64, rectangle_length: i64| rectangle_width * rectangle_length;

    // ? anonymous function expression with closure syntax />


    
    // ? < Passing functions as arguments to other functions

    say_hello_v1(say_how_are_you_v1);

    say_hello_v1(|| println!("how are you?"));

    let say_hello_v2 = |callback_function: &dyn Fn()| {
        println!("hello");
        callback_function();
    };

    let say_how_are_you_v2 = || println!("how are you?");

    say_hello_v2(&say_how_are_you_v2);

    say_hello_v2(&|| println!("how are you?"));

    // ? Passing functions as arguments to other functions />



    // ? < Assigning functions to variables, but unfortunately we can't freely storing them in data structures

    let get_rectangle_area_v2_copy = |rectangle_width: i64, rectangle_length: i64| -> i64 {
        return rectangle_width * rectangle_length;
    };

    let get_rectangle_area_v3_copy = |rectangle_width: i64, rectangle_length: i64| rectangle_width * rectangle_length;

    // ? Assigning functions to variables, but unfortunately we can't freely storing them in data structures />



    // ? < Returning functions as values from other functions

    let multiply_by_2 = multiply_v1(2);
    let multiply_by_2_result = multiply_by_2(10); // 20
    println!("{:?}", multiply_by_2_result);

    // ? Returning functions as values from other functions />
}

File: codes/gnu_octave/007_functions.m
% ? function statement or function declaration

function getrectangleareav1result = getrectangleareav1(rectanglewidth, rectanglelength)
    getrectangleareav1result = (rectanglewidth * rectanglelength);
end
% call the function example: getrectangleareav1(7, 5)

% ? function expression

getrectangleareav2 = @(rectanglewidth, rectanglelength) (rectanglewidth * rectanglelength);
% call the function example: getrectangleareav2(7, 5)

% ? anonymous function

@(rectanglewidth, rectanglelength) (rectanglewidth * rectanglelength);

% ? Passing functions as arguments to other functions

function sayhello(callbackfunction)
    disp("hello");
    callbackfunction();
end

function sayhowareyou()
    disp("how are you?");
end

sayhello(@sayhowareyou);

sayhello(@() disp("how are you?"));

% ? Assigning functions to variables or storing them in data structures

getrectangleareav2 = @(rectanglewidth, rectanglelength) (rectanglewidth * rectanglelength);

myarrayofgetrectangleareafunctions = { ...
    @getrectangleareav1, ...
    getrectangleareav2, ...
    @(rectanglewidth, rectanglelength) (rectanglewidth * rectanglelength) ...
};
getrectangleareafunctionresult1 = myarrayofgetrectangleareafunctions{1}(7, 5);
getrectangleareafunctionresult2 = myarrayofgetrectangleareafunctions{2}(7, 5);
getrectangleareafunctionresult3 = myarrayofgetrectangleareafunctions{3}(7, 5);

function exponentiationresult = exponentiation(a, b)
    exponentiationresult = (a ^ b);
end

simplecalculator = struct( ...
    "exponentiation", {@exponentiation}, ...
    "multiplication", {@(a, b) (a * b)} ...
);
simplecalculatorresult1 = simplecalculator.("exponentiation")(2, 4);
simplecalculatorresult2 = simplecalculator.("multiplication")(7, 5);

function multiplyresult = multiplyv1(a)
    function multiplybyresult = multiplyby(b)
        multiplybyresult = (a * b);
    end
    multiplyresult = @multiplyby;
end
multiplyby2 = multiplyv1(2);
multiplyby2result = multiplyby2(10); % 20

function multiplyresult = multiplyv2(a)
    multiplyresult = @(b) (a * b);
end
multiplyby3 = multiplyv2(3);
multiplyby3result = multiplyby3(10); % 30

multiplyv3 = @(a) (@(b) (a * b));
multiplyby4 = multiplyv3(4);
multiplyby4result = multiplyby4(10); % 40
